src/zkevm_specs/state.py

```diff
@@ -263,6 +265,68 @@ def check_account_destructed(row: Row, row_prev: Row):
     # TODO: Missing constraints
 
 
+@is_circuit_code
+def check_tx_log(row: Row, row_prev: Row):
+    # tx_id | log_id | field_tag | index | value
+    tx_id = row.keys[1]
+    pre_tx_id = row_prev.keys[1]
+    log_id = row.keys[2]
+    pre_field_tag = row_prev.keys[3]
+    field_tag = row.keys[3]
+    index = row.keys[4]
+    pre_index = row_prev.keys[4]
+
+    # is_write is always true
+    assert row.is_write == 1
+    # reset log_id when tx_id increases
+    if row.tag() == row_prev.tag():
+        if tx_id != pre_tx_id:
+            assert tx_id == pre_tx_id + 1
+            assert log_id == 0
+        else:
+            # increase log_id when tag changes to Address, make sure tag can only increase(Non-Decreasing),
+            # when log_index stays same,
+            if pre_field_tag == U256(TxLogFieldTag.Address):
+                assert (field_tag - pre_field_tag).n > 0
+            # make sure if tag Data appear, data_index can only increase by one when tag stays same.
+            # make sure if tag Topic appear, topic_index in range [0,4),can only increase by one when tag stays same.
+            if field_tag == U256(TxLogFieldTag.Topic):
+                assert_in_range(index, 0, 3)
+            if field_tag not in [U256(TxLogFieldTag.Topic), U256(TxLogFieldTag.Data)]:
+                assert index == 0
+            elif pre_field_tag == field_tag:
+                assert index == pre_index + 1
+
+        # both log_index topic_index data_index are non negative, all starts with zero
+    assert index.n >= 0 and log_id.n >= 0
+
+    # make sure if log set is not empty within receipt/tx, tx must be successful status because failed execution will revert the log data,
+    # in other words, logs section must be empty list for failed tx
```

Comment:
This comment doesn't contain any constraint.  How is this enforced?

Maybe we do need to track the log_id in the EVM circuit after all, and at the end of the transaction assert that the number of logs counted in the EVM matches the last log_id for the tx in the rw_table (this would need a flag on the last log of the tx in the rw_table).

This is similar to the check we will need to do for the rwc, to assert that no malicious entries are inserted in the rw_table that are not looked up by the EVM circuit.

Reply:
Good topic,  I am thinking of adding another tag into TxReceiptFieldTag from receipt RLP encoding perspective, because it is tx/receipt instead of log specific :
`class TxReceiptFieldTag(IntEnum):
 
    # The following are write-only data inside a transaction, they will be written in
    # State circuit directly.
    PostStateOrStatus = auto()  # flag indicates whether if a tx succeed or not
    # the cumulative gas used in the block containing the transaction receipt as of immediately
    # after the transaction has happened.
    CumulativeGasUsed = auto()
    # record how many log entries in the receipt/tx 
    LogLength = auto()`. 

`LogLength` represents how many log entries generated in the receipt/tx , it should be required by receipt RLP encoding , we can check this value match within `end_tx` step's  `log_id` state to prevent malicious insertion.  
let me know what your think !

Reply:
Adding the `LogLength` tag sounds good to me.  I'm assuming that you'd have the following constraints:
1. `LogLength` can only appear once for every `tx_id`
2. `LogLength` must be the last row for a given `tx_id`
3. Either the `LogLength` value is the previous row's `log_id` or 0.

Reply:
the first two are good to me, for the third one, a bit change: due to LogLength is belong to  `TxReceipt` tag not `TxLog` tag, these two tag's rows are not adjacent,. this check will conduct in `end_tx` evm  step gadget .  `end_tx` gadget knows the last `log_id` and can do lookup check.

Reply:
added `LogLength` in [7208290](https://github.com/appliedzkp/zkevm-specs/pull/159/commits/72082901f7731cff2fe9eeb0f8ecc4833adb8f78)